### PR TITLE
Perf decode string

### PR DESCRIPTION
**Pull request by zonghaishang**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #https://github.com/apache/dubbo-go-hessian2/issues/186

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

优化思路：

1.  直接使用utf-8 byte解码，性能最高；之前先解码成 rune, 对rune解码成string，及其耗费性能
2.  增加批量string chunk copy, 降低read调用
3.  使用unsafe转换string

### CODE REVIEW

File: string.go [hunk 1]
```diff
@@ -343,36 +342,137 @@ func (d *Decoder) decString(flag int32) (string, error) {
...
+						bytesBuf = bs
+					}
 				default:
 					return s, perrors.New("expect string tag")
 				}
 			}
 
-			r, charLen, byteLen, err = decodeUcs4Rune(d.reader)
+			if buffed := d.Buffered(); chunkLen > 0 {
+				nread, err := d.next(bytesBuf[offset : offset+chunkLen])
```
Comment (gaoxinge) [inline string.go:373]: 我的理解：`chunkLen`是表示字符串的长度，不是表示字符串在对应编码下byte的长度。byte的长度应该比`chunkLen`大才对。这样不会有问题么？

Comment (zonghaishang) [reply string.go:373]: 不会有问题，chunkLen是char的长度，优化后的版本会检查后续剩余字节。这里只不过优先一次性读取chunkLen字节长度而已。

File: decode_test.go [hunk 2]
```diff
@@ -126,3 +127,16 @@ func testDecodeFrameworkFunc(t *testing.T, method string, expected func(interfac
 	}
 	expected(r)
 }
+
+func BenchmarkDecodeStringOptimized(t *testing.B) {
```
[the remark this answers is not in the document]
Comment (zonghaishang) [reply decode_test.go]: ok, fixed.

File: decode.go [hunk 3]
```diff
@@ -50,15 +50,50 @@ func NewDecoder(b []byte) *Decoder {
...
 // utilities
 /////////////////////////////////////////
 
+func (d *Decoder) Reset(b []byte) *Decoder {
+	// reuse reader buf, avoid allocate
+	d.reader.Reset(bytes.NewReader(b))
+	d.typeRefs = &TypeRefs{records: map[string]bool{}}
+
+	if d.refs != nil || d.classInfoList != nil {
+		d.refs, d.classInfoList = nil, nil
```
Comment (AlexStocks) [inline decode.go:92]: pls split it into two lines. we do not need such not easily understandable codes.

Comment (zonghaishang) [reply decode.go]: ok, fixed.

File: decode.go [hunk 4]
```diff
@@ -50,15 +50,54 @@ func NewDecoder(b []byte) *Decoder {
...
 /////////////////////////////////////////
 // utilities
 /////////////////////////////////////////
 
+func (d *Decoder) Reset(b []byte) *Decoder {
+	// reuse reader buf, avoid allocate
+	d.reader.Reset(bytes.NewReader(b))
+	d.typeRefs = &TypeRefs{records: map[string]bool{}}
+
+	if d.refs != nil {
```
Comment (AlexStocks) [inline decode.go:91]: the two following lines is enough. u do not need write two if clauses.
```
d.refs = nil
d.classInfoList = nil
```

File: string_test.go [hunk 5]
```diff
@@ -158,11 +158,25 @@ func TestStringEncode(t *testing.T) {
...
+	}
+}
+
 func TestStringEmoji(t *testing.T) {
 	// see: test_hessian/src/main/java/test/TestString.java
 	s0 := "emoji🤣"
 	s0 += ",max" + string(rune(0x10FFFF))
 
-	testDecodeFramework(t, "customReplyStringEmoji", s0)
+	// todo 这里正确拿到hessian解码字节数组，但是构造string的时候，不是rune，emoji表情符号显示有点问题，修改assert？？？
```
Comment (AlexStocks) [inline string_test.go:179]: u can modify the assert.

File: decode.go (hunk 4 above)
Comment (zonghaishang) [reply decode.go:91]: if d.refs == nill or d.classInfoList == nil  already,  do nothing.

File: string.go [hunk 6]
```diff
@@ -343,36 +342,139 @@ func (d *Decoder) decString(flag int32) (string, error) {
...
-					runeData = bs
 
+					if chunkLen < 0 {
+						chunkLen = 0
+					}
+					if charLen < 0 {
+						charLen = 0
+					}
+
+					chunkLen += charLen
```
Comment (gaoxinge) [inline string.go:357]: 直接写`chunkLen=charLen`就行了。因为从上面的分析看，此时的`chunkLen`一定等于0.

File: string.go [hunk 7]
```diff
@@ -343,36 +342,139 @@ func (d *Decoder) decString(flag int32) (string, error) {
...
+						return *(*string)(unsafe.Pointer(&b)), nil
+					}
+
+					// we need to detect next chunk
+					continue
+				}
+			}
+
+			// decode byte
+			ch, err := d.readByte()
```
Comment (gaoxinge) [inline string.go:434]: 为啥最后要decode一下？上面的循环解析cover不掉什么corner case么？能解释一下么？

File: string.go (hunk 1 above)
Comment (gaoxinge) [reply string.go:371]: 嗯，我懂了。

File: string.go (hunk 7 above)
Comment (zonghaishang) [reply string.go:434]: 两个目的：
1. 触发buffer fill data.
2. 前面只能保证读取chunkLen字节数, 并且读取整个char的utf-8编码字节(保证通过char的字节不被拆分)，但是可能并没有处理所有chunkLen个数的字符。后续的read可以处理完所有chunkLen字符。

**Comment by zonghaishang** [comment]
> this PR will let the hessian only support at most three bytes length utf8, while four bytes mathematical symbols, and emoji will not supprted. even more the max length of a utf8 character can be six. And the worst is the length of string in hessian definition is the length of A 16-bit unicode character string encoded in UTF-8.
> ref
> 
> * https://en.wikipedia.org/wiki/UTF-8
> * http://hessian.caucho.com/doc/hessian-serialization.html##string
> * #131
@wongoo 
Any Suggestions for fixing it? I've tried parsing 4-byte utf-8 before, but Java doesn't write 4-byte utf-8

**Comment by AlexStocks** [comment]
> > this PR will let the hessian only support at most three bytes length utf8, while four bytes mathematical symbols, and emoji will not supprted. even more the max length of a utf8 character can be six. And the worst is the length of string in hessian definition is the length of A 16-bit unicode character string encoded in UTF-8.
> > ref
> > 
> > * https://en.wikipedia.org/wiki/UTF-8
> > * http://hessian.caucho.com/doc/hessian-serialization.html##string
> > * #131
> >   @wongoo
> >   Any Suggestions for fixing it? I've tried parsing 4-byte utf-8 before, but Java doesn't write 4-byte utf-8

maybe u can set an issue for dubbo hessian2.

**Comment by zonghaishang** [comment]
I'll explore the byte code for the go emoji first, and there may be a way to fix it

**Comment by AlexStocks** [comment]
> I'll explore the byte code for the go emoji first, and there may be a way to fix it

glad to hear that.

File: string.go [hunk 8]
```diff
@@ -343,36 +342,139 @@ func (d *Decoder) decString(flag int32) (string, error) {
...
+				if err != nil {
+					if err == io.EOF {
+						break
+					}
+					return s, perrors.WithStack(err)
+				}
+
+				// quickly detect the actual number of bytes
+				prev := offset
+				for i, len := offset, offset+nread; i < len; chunkLen-- {
```
[the remark this answers is not in the document]
Comment (zonghaishang) [reply string.go]: optimized.

File: string.go [hunk 9]
```diff
@@ -343,36 +342,139 @@ func (d *Decoder) decString(flag int32) (string, error) {
...
+
+						if err != nil {
+							return s, perrors.WithStack(err)
+						}
+					}
+				}
+
+				// the expected length string has been processed.
+				if chunkLen <= 0 {
+					if last {
```
[the remark this answers is not in the document]
Comment (zonghaishang) [reply string.go]: removed already.

**Comment by wongoo** [comment]
in fact, when decoding you can't know how many bytes you should read, until you analysis all bytes. A possible improvement may be you can first read a chunk length of bytes first, and then loop analysis and read more until reach the length of chunk. This may be useful to improve performance for most situation.

**Comment by zonghaishang** [comment]
> in fact, when decoding you can't know how many bytes you should read, until you analysis all bytes. A possible improvement may be you can first read a chunk length of bytes first, and then loop analysis and read more until reach the length of chunk. This may be useful to improve performance for most situation.

The current optimization already includes this idea.

File: string_test.go (hunk 5 above)
Comment (zonghaishang) [reply string_test.go:179]: Support emoji decoding, performance reduced by 4%， :-(

Comment (AlexStocks) [reply string_test.go:179]: but we should support it.

**Comment by zonghaishang** [comment]
@wongoo  please take a look.

**Comment by zonghaishang** [comment]
2 votes support, merge to master now.

File: string_test.go [hunk 10]
```diff
@@ -158,11 +158,25 @@ func TestStringEncode(t *testing.T) {
...
+		d.Reset(buf)
+	}
+}
+
 func TestStringEmoji(t *testing.T) {
 	// see: test_hessian/src/main/java/test/TestString.java
 	s0 := "emoji🤣"
 	s0 += ",max" + string(rune(0x10FFFF))
 
+	// todo 这里正确拿到hessian解码字节数组，但是构造string的时候，不是rune，emoji表情符号显示有点问题，修改assert？？？
```
Comment (wongoo) [inline string_test.go:179]: remove the comment

Comment (zonghaishang) [reply string_test.go:179]: ok, 我单独提个pr

Comment (AlexStocks) [reply string_test.go:179]: @wongoo are there any other problems? If not, we can reserve this comment in my opinion. @zonghaishang , u can translate this comment into english.

**Pull request merged**